Failure in SystemIndexMigrationIT#testSystemIndexMigrationCanBeInterruptedWithShutdown

### CI Link

https://gradle-enterprise.elastic.co/s/xe5tsxwd6ky64

### Repro line

./gradlew ':modules:reindex:internalClusterTest' --tests "org.elasticsearch.migration.SystemIndexMigrationIT.testSystemIndexMigrationCanBeInterruptedWithShutdown" -Dtests.seed=21ABF49FDF24F147 -Dtests.locale=ar-DZ -Dtests.timezone=Europe/Vatican -Druntime.java=17

### Does it reproduce?

No

### Applicable branches

8.0

### Failure history

Looks like the first time it has failed

### Failure excerpt

```
org.elasticsearch.migration.SystemIndexMigrationIT > testSystemIndexMigrationCanBeInterruptedWithShutdown FAILED
    java.lang.AssertionError: expected migration to fail due to restarting only data node
    Expected: <ERROR>
         but: was <NO_MIGRATION_NEEDED>
        at __randomizedtesting.SeedInfo.seed([21ABF49FDF24F147:5E92BED7875328A5]:0)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.junit.Assert.assertThat(Assert.java:956)
        at org.elasticsearch.migration.SystemIndexMigrationIT.lambda$testSystemIndexMigrationCanBeInterruptedWithShutdown$1(SystemIndexMigrationIT.java:119)
        at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:1069)
        at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:1042)
        at org.elasticsearch.migration.SystemIndexMigrationIT.testSystemIndexMigrationCanBeInterruptedWithShutdown(SystemIndexMigrationIT.java:110)
```

I think I see the problem with the test. 

The test submits a system index migration task and then it restarts the master node. After restarting the master node, the tests checks if the system index migration task is in an error state. However, there's a small time window where the test wouldn't work as expected. 

We submit the system index migration task on the transport thread asynchronously, and then we restart the master. What did happen in this instance is that the master had already gone down when the transport thread was about to start the task, so we simply got node closed exception, and migration never started:

```
1> Caused by: org.elasticsearch.node.NodeClosedException: node closed {node_t1}{oWFG_T2dT0alL1om9I2-nw}{_AP13LVIRhS5G8wlEsjD1w}{127.0.0.1}{127.0.0.1:42755}{cdfhilmrstw} |  
-- | --
  | 1> 	at org.elasticsearch.transport.TransportService.sendRequestInternal(TransportService.java:835) ~[elasticsearch-8.0.2-SNAPSHOT.jar:8.0.2-SNAPSHOT] |  
  | 1> 	... 12 more
```

The test waits to see if the task has started before it continues with the node shutdown, by checking if the task is in the cluster state list. I think we need a bit stronger test here, we need to ensure the migration has started by checking if the task in the list has a state object as well.

I'll make a PR with this proposed change.

--- a/modules/reindex/src/internalClusterTest/java/org/elasticsearch/migration/SystemIndexMigrationIT.java
+++ b/modules/reindex/src/internalClusterTest/java/org/elasticsearch/migration/SystemIndexMigrationIT.java
@@ -75,8 +75,13 @@ public class SystemIndexMigrationIT extends AbstractFeatureMigrationIntegTest {
         createSystemIndexForDescriptor(INTERNAL_MANAGED);
 
         final ClusterStateListener clusterStateListener = event -> {
+            PersistentTasksCustomMetadata.PersistentTask<?> task = PersistentTasksCustomMetadata.getTaskWithId(
+                event.state(),
+                SYSTEM_INDEX_UPGRADE_TASK_NAME
+            );
 
-            if (PersistentTasksCustomMetadata.getTaskWithId(event.state(), SYSTEM_INDEX_UPGRADE_TASK_NAME) != null
+            if (task != null
+                && task.getState() != null // Make sure the task is really started
                 && hasBlocked.compareAndSet(false, true)) {
                 try {
                     logger.info("Task created");

